nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/DecryptMode.groovy

```diff
@@ -0,0 +1,322 @@
+            printUsageAndExit(e.getMessage(), EncryptConfigMain.EXIT_STATUS_FAILURE)
+        }
+    }
+
+    void run(DecryptConfiguration config) throws Exception {
+
+        if (!config.fileType) {
+
+            // Try to load the input file to auto-detect the file type
+            boolean isPropertiesFile = PropertiesEncryptor.supportsFile(config.inputFilePath)
+
+            boolean isXmlFile = XmlEncryptor.supportsFile(config.inputFilePath)
+
+            if (ToolUtilities.isExactlyOneTrue(isPropertiesFile, isXmlFile)) {
+                if (isPropertiesFile) {
+                    config.fileType = FileType.properties
+                    logger.debug("Auto-detection of input file type determined the type to be: ${FileType.properties}")
+                }
+                if (isXmlFile) {
+                    config.fileType = FileType.xml
+                    logger.debug("Auto-detection of input file type determined the type to be: ${FileType.xml}")
+                }
+            }
+
+            // Could we successfully auto-detect?
+            if (!config.fileType) {
+                throw new RuntimeException("Auto-detection of input file type failed. Please re-run the tool specifying the file type with the -t/--fileType flag.")
+            }
+        }
+
+        String decryptedSerializedContent = null
+        switch (config.fileType) {
+
+            case FileType.properties:
+                PropertiesEncryptor propertiesEncryptor = new PropertiesEncryptor(null, config.decryptionProvider)
+                Properties properties = propertiesEncryptor.loadFile(config.inputFilePath)
+                properties = propertiesEncryptor.decrypt(properties)
+                decryptedSerializedContent = propertiesEncryptor.serializePropertiesAndPreserveFormatIfPossible(properties, config.inputFilePath)
+                break
+
+            case FileType.xml:
+                XmlEncryptor xmlEncryptor = new XmlEncryptor(null, config.decryptionProvider) {
+                    @Override
+                    List<String> serializeXmlContentAndPreserveFormat(String updatedXmlContent, File originalInputXmlFile) {
+                        // For decrypting unknown, generic XML, this tool will not support preserving the format
+                        return updatedXmlContent.split("\n")
+                    }
+                }
+
+                String xmlContent = xmlEncryptor.loadXmlFile(config.inputFilePath)
+                xmlContent = xmlEncryptor.decrypt(xmlContent)
+                decryptedSerializedContent = xmlEncryptor.serializeXmlContentAndPreserveFormatIfPossible(xmlContent, config.inputFilePath)
+                break
+
+            default:
+                throw new RuntimeException("Unsupported file type '${config.fileType}'")
+        }
+
+        if (!decryptedSerializedContent) {
+            throw new RuntimeException("Failed to load and decrypt input file.")
+        }
+
+        if (config.outputToFile) {
```

Comment:
I don't think `decrypt` should allow output to a file. At this stage, it is safer to require the output goes to STDOUT to avoid decrypting files in place.

Reply:
Now that I think about it, I agree with you. That would be the better/safer approach for this initial offering of the decrypt capability. Will update